Improve documentation for std.typecons.Proxy

MetaLang opened the pull request:
- Add a better opening explanation
- Add Params section
- Add documented unittest examples

Hackerpilot commented on std/typecons.d:
```diff
@@ -4399,7 +4399,19 @@ unittest
-Make proxy for $(D a).
+    Creates a proxy for the value `a` that will forward all operations
+    while disabling implicit conversions. The aliased item `a` must be 
+    an $(B lvalue). This is useful for creating a new type from the 
```
Once I fix https://issues.dlang.org/show_bug.cgi?id=14328, these should be turned into links.

CyberShadow commented on std/typecons.d:
```diff
@@ -4628,6 +4640,63 @@ unittest
+    {
+        //Ok, "new Object" is an lvalue.
+        //This code is a bit weird, though.
+        mixin Proxy!(new Object);
```
What's the point of this example?

MetaLang replied:
To demonstrate that *any* lvalue can be used with Proxy, even if it doesn't make sense.

CyberShadow commented on std/typecons.d:
```diff
@@ -4628,6 +4640,63 @@ unittest
+/**
+    There is one exception to the fact that the new type is not related to the
+    old type. Pseudo member functions that use 
+    $(LINK2 http://ddili.org/ders/d.en/ufcs.html, Universal Function Call Syntax)
```
There's no reason to link to third-party resources when the official ones suffice.

http://dlang.org/function.html#pseudo-member

MetaLang replied:
Thanks, I was looking for that and couldn't find it.

MetaLang commented:
I have changed the link explaining UFCS to http://dlang.org/function.html#pseudo-member and changed the documented unittest to use actual useful code.

MetaLang commented:
Ping

quickfur commented on std/typecons.d:
```diff
@@ -4628,6 +4640,65 @@ unittest
+    NewIntType nit = 0;
+    nit++;
+    assert(nit == 1);
+    
```
What about showing that the proxied type is not implicitly convertible to/from the base value, as well as constitute a distinct type?

MetaLang replied:
That's covered by the unit test above this one.

quickfur replied:
Ah I see, thanks.

quickfur commented:
Let's merge. Further improvements can be done in a followup PR.

quickfur commented:
Auto-merge toggled on
